crates/dbsp/src/circuit/circuit_builder.rs

```diff
@@ -564,327 +574,978 @@ impl Edge {
+    /// using the default scheduler (currently [`DynamicScheduler`]).
+    ///
+    /// Returns a handle to the newly-created circuit and a user-defined value
+    /// returned by the constructor.  This value typically contains one or more
+    /// input handles used to push data to the circuit at runtime (see
+    /// [`RootCircuit::add_input_stream`], [`RootCircuit::add_input_zset`], and
+    /// related methods).
+    ///
+    /// The circuit created using this API does
+    /// not have its own worker thread(s) and will instead execute in the
+    /// context of the calling thread.  The caller should use the returned
+    /// [`CircuitHandle`] to control the execution of the circuit.
+    // TODO: Allow `constructor` to fail.
+    pub fn build<F, T>(constructor: F) -> Result<(CircuitHandle, T), SchedulerError>
+    where
+        F: FnOnce(&mut RootCircuit) -> T,
+    {
+        Self::build_with_scheduler::<F, T, DynamicScheduler>(constructor)
+    }
+
+    /// Create a circuit and prepare it for execution.
+    ///
+    /// Similar to [`build`](`Self::build`), but with a user-specified
+    /// [`Scheduler`] implementation.
+    pub fn build_with_scheduler<F, T, S>(
+        constructor: F,
+    ) -> Result<(CircuitHandle, T), SchedulerError>
+    where
+        F: FnOnce(&mut RootCircuit) -> T,
+        S: Scheduler + 'static,
+    {
+        let mut circuit = RootCircuit::new();
+        let res = constructor(&mut circuit);
+        let executor =
+            Box::new(<OnceExecutor<S>>::new(&circuit)?) as Box<dyn Executor<RootCircuit>>;
+
+        // Alternatively, `CircuitHandle` should expose `clock_start` and `clock_end`
+        // APIs, so that the user can reset the circuit at runtime and start
+        // evaluation from clean state without having to rebuild it from
+        // scratch.
+        circuit.log_scheduler_event(&SchedulerEvent::clock_start());
+        circuit.clock_start(0);
+        Ok((CircuitHandle { circuit, executor }, res))
+    }
+}
+
+impl RootCircuit {
+    // Create new top-level circuit.  Clients invoke this via the
+    // [`RootCircuit::build`] API.
+    fn new() -> Self {
+        Self {
+            inner: Rc::new(RefCell::new(CircuitInner::new(
+                (),
+                0,
+                NodeId::root(),
+                GlobalNodeId::root(),
+                Rc::new(RefCell::new(HashMap::new())),
+                Rc::new(RefCell::new(HashMap::new())),
+            ))),
+            time: Rc::new(RefCell::new(())),
+        }
+    }
+}
 
-        // Safe because the circuit isn't running yet.
-        unsafe { (*stream.val.get()).consumers += 1 };
+impl RootCircuit {
+    /// Attach a circuit event handler to the top-level circuit (see
+    /// [`super::trace::CircuitEvent`] for a description of circuit events).
+    ///
+    /// This method should normally be called inside the closure passed to
+    /// [`RootCircuit::build`] before adding any operators to the circuit, so
```

Comment:
"events" are just construction events?
Maybe they should be called construction_event

Reply:
They are called circuit events as opposed to scheduler events that happen at runtime.  I agree the naming is unfortunate, but it's not really something that changed in this PR, so I'd rather not fix them here.